src/main/groovy/liquibase/sdk/test/change/ChangeObjectTests.groovy

```diff
@@ -0,0 +1,153 @@
+import liquibase.sdk.test.util.TestUtils
+import liquibase.util.StringUtil
+import org.skyscreamer.jsonassert.JSONAssert
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import java.util.logging.Logger
+
+class ChangeObjectTests extends Specification {
+
+    final static List supportedChangeLogFormats = ['xml', 'sql', 'json', 'yml', 'yaml'].asImmutable()
+
+    @Unroll
+    def "apply #testInput.changeObject against #testInput.databaseName; verify generated SQL and DB snapshot"() {
+        given:
+        Liquibase liquibase = TestUtils.createLiquibase(testInput.pathToChangeLogFile, testInput.database)
+
+        String expectedSql = cleanSql(FileUtils.getFileContent(testInput.changeObject, testInput.database.shortName, testInput.database.databaseMajorVersion, testInput.database.databaseMinorVersion, "liquibase/sdk/test/change/expectedSql", ".sql"))
+        String expectedSnapshot = FileUtils.getFileContent(testInput.changeObject, testInput.database.shortName, testInput.database.databaseMajorVersion, testInput.database.databaseMinorVersion, "liquibase/sdk/test/change/expectedSnapshot", ".json")
+        List<CatalogAndSchema> catalogAndSchemaList = TestUtils.getCatalogAndSchema(testInput.database, testInput.dbSchema)
+
+        when:
+        def generatedSql = cleanSql(TestUtils.toSqlFromLiquibaseChangeSets(liquibase))
+
+        then:
+        assert expectedSnapshot != null: "No expectedSnapshot for ${testInput.changeObject} against ${testInput.database.shortName} ${testInput.database.databaseMajorVersion}.${testInput.database.databaseMinorVersion}"
+
+        if (expectedSql != null && !testInput.pathToChangeLogFile.endsWith(".sql")) {
+            assert generatedSql == expectedSql: "Expected SQL does not match actual sql. Deleting the existing expectedSql file will test that the new SQL works correctly and will auto-generate a new version if it passes"
+            if (!TestConfig.instance.revalidateSql) {
+                return //sql is right. Nothing more to test
+            }
+        }
+
+        assert testInput.database.getConnection() instanceof JdbcConnection: "We cannot verify the following SQL works works because the database is offline:\n${generatedSql}"
+
+        when:
+        liquibase.update(testInput.context)
+
+        String jsonSnapshot = SnapshotHelpers.getJsonSnapshot(testInput.database, catalogAndSchemaList)
+        liquibase.rollback(liquibase.databaseChangeLog.changeSets.size(), testInput.context)
+
+        then:
+        snapshotMatchesSpecifiedStructure(expectedSnapshot, jsonSnapshot)
+
+        if (expectedSql == null && !testInput.pathToChangeLogFile.endsWith(".sql")) {
+            //save generated sql as expected sql for future runs
+            saveAsExpectedSql(generatedSql, testInput)
+        }
+
+        where:
+        testInput << buildTestInput()
+    }
+
+    /**
+     * Standardizes sql content. Removes line ending differences, and unnessisary leading/trailing whitespace
```

Comment:
I added ChangeObjectTestHelper to move all util methods there and make test class itself more readable.